File: xNet/src/main/java/org/sqlunet/model/TreeFactory.java
/*
 * Copyright (c) 2023. Bernard Bou
 */

package org.sqlunet.model;

import android.os.Handler;
import android.os.Looper;

import org.sqlunet.treeview.control.ColdQueryController;
import org.sqlunet.treeview.control.CompositeValue;
import org.sqlunet.treeview.control.HotQueryController;
import org.sqlunet.treeview.control.IconTextController;
import org.sqlunet.treeview.control.LeafController;
import org.sqlunet.treeview.control.Link;
import org.sqlunet.treeview.control.LinkHotQueryController;
import org.sqlunet.treeview.control.LinkLeafController;
import org.sqlunet.treeview.control.LinkNodeController;
import org.sqlunet.treeview.control.LinkQueryController;
import org.sqlunet.treeview.control.LinkTreeController;
import org.sqlunet.treeview.control.MoreController;
import org.sqlunet.treeview.control.Query;
import org.sqlunet.treeview.control.TextController;
import org.sqlunet.treeview.control.TreeController;
import org.sqlunet.treeview.model.TreeNode;

import androidx.annotation.DrawableRes;
import androidx.annotation.NonNull;

/**
 * Tree factory
 *
 * @author <a href="mailto:1313ou@gmail.com">Bernard Bou</a>
 */
public class TreeFactory
{
	// private static final String TAG = "TreeFactory";

	// NON-TREE (without tree junction icon)

	/**
	 * Make text node
	 *
	 * @param value       character sequence
	 * @param breakExpand break expand flag
	 * @return created node
	 */
	@NonNull
	static public TreeNode makeTextNode(@NonNull final CharSequence value, final boolean breakExpand)
	{
		return new TreeNode(value, new TextController(breakExpand));
	}

	/**
	 * Make icon-text node
	 *
	 * @param text        text
	 * @param icon        icon
	 * @param breakExpand break expand flag
	 * @return created node
	 */
	@NonNull
	static public TreeNode makeIconTextNode(@NonNull final CharSequence text, final int icon, final boolean breakExpand)
	{
		return new TreeNode(new CompositeValue(text, icon), new IconTextController(breakExpand), false);
	}

	/**
	 * Make leaf node
	 *
	 * @param text        text
	 * @param icon        icon (extra icon after tree icon)
	 * @param breakExpand break expand flag
	 * @return created node
	 */
	@NonNull
	static public TreeNode makeLeafNode(@NonNull final CharSequence text, final int icon, final boolean breakExpand)
	{
		return new TreeNode(new CompositeValue(text, icon), new LeafController(breakExpand), false);
	}

	/**
	 * Make more (leaf) node
	 *
	 * @param text        text
	 * @param icon        icon (extra icon after tree icon)
	 * @param breakExpand break expand flag
	 * @return created node
	 */
	@NonNull
	static public TreeNode makeMoreNode(@NonNull final CharSequence text, final int icon, final boolean breakExpand)
	{
		return new TreeNode(new CompositeValue(text, icon), new MoreController(breakExpand), false);
	}

	/**
	 * Make icon-text-link node
	 *
	 * @param text        text
	 * @param icon        icon
	 * @param link        link
	 * @param breakExpand break expand flag
	 * @return created node
	 */
	@NonNull
	static public TreeNode makeLinkNode(@NonNull final CharSequence text, final int icon, final boolean breakExpand, final Link link)
	{
		return new TreeNode(new CompositeValue(text, icon, link), new LinkNodeController(breakExpand), false);
	}

	/**
	 * Make link leaf node
	 *
	 * @param text        text
	 * @param icon        icon
	 * @param breakExpand break expand flag
	 * @param link        link
	 * @return created node
	 */
	@NonNull
	static public TreeNode makeLinkLeafNode(@NonNull final CharSequence text, final int icon, final boolean breakExpand, final Link link)
	{
		return new TreeNode(new CompositeValue(text, icon, link), new LinkLeafController(breakExpand));
	}

	// TREE

	/**
	 * Add tree node
	 *
	 * @param value       character sequence
	 * @param icon        icon resource id
	 * @param breakExpand break expand flag
	 * @return created node
	 */
	@NonNull
	static public TreeNode makeTreeNode(@NonNull final CharSequence value, final int icon, final boolean breakExpand)
	{
		return new TreeNode(new CompositeValue(value, icon), new TreeController(breakExpand));
	}

	/**
	 * Make link tree node
	 *
	 * @param text        text
	 * @param icon        icon
	 * @param breakExpand break expand flag
	 * @param link        link
	 * @return created node
	 */
	@NonNull
	static public TreeNode makeLinkTreeNode(@NonNull final CharSequence text, final int icon, final boolean breakExpand, final Link link)
	{
		return new TreeNode(new CompositeValue(text, icon, link), new LinkTreeController(breakExpand));
	}

	/**
	 * Make query node
	 *
	 * @param text        label text
	 * @param icon        icon
	 * @param breakExpand break expand flag
	 * @param query       query
	 * @return created node
	 */
	@NonNull
	static public TreeNode makeQueryNode(@NonNull final CharSequence text, final int icon, final boolean breakExpand, final Query query)
	{
		return new TreeNode(new CompositeValue(text, icon, query), new ColdQueryController(breakExpand));
	}

	/**
	 * Make hot (self-triggered) query node
	 *
	 * @param text        label text
	 * @param icon        icon
	 * @param breakExpand break expand flag
	 * @param query       query
	 * @return created node
	 */
	@NonNull
	static public TreeNode makeHotQueryNode(@NonNull final CharSequence text, final int icon, final boolean breakExpand, final Query query)
	{
		final HotQueryController controller = new HotQueryController(breakExpand);
		final TreeNode result = new TreeNode(new CompositeValue(text, icon, query), controller);

		final Handler handler = new Handler(Looper.getMainLooper());
		handler.post(controller::processQuery);

		return result;
	}

	/**
	 * Make hot (self-triggered) query node
	 *
	 * @param text           label text
	 * @param icon           icon
	 * @param breakExpand    break expand flag
	 * @param query          query
	 * @param link           link
	 * @param buttonImageRes image drawable id for button, 0 for default
	 * @return created node
	 */
	@NonNull
	static public TreeNode makeLinkHotQueryNode(@NonNull final CharSequence text, @DrawableRes final int icon, final boolean breakExpand, final Query query, final Link link, @DrawableRes final int buttonImageRes)
	{
		final HotQueryController controller = new LinkHotQueryController(breakExpand, buttonImageRes);
		final TreeNode result = new TreeNode(new CompositeValue(text, icon, query, link), controller);

		final Handler handler = new Handler(Looper.getMainLooper());
		handler.post(controller::processQuery);

		return result;
	}

	/**
	 * Make link query node
	 *
	 * @param text           label text
	 * @param icon           icon
	 * @param breakExpand    break expand flag
	 * @param query          query
	 * @param link           link
	 * @param buttonImageRes image drawable id for button, 0 for default
	 * @return created node
	 */
	@NonNull
	static public TreeNode makeLinkQueryNode(@NonNull final CharSequence text, @DrawableRes final int icon, final boolean breakExpand, final Query query, final Link link, @DrawableRes final int buttonImageRes)
	{
		return new TreeNode(new CompositeValue(text, icon, query, link), new LinkQueryController(breakExpand, buttonImageRes));
	}

	// H E L P E R S

	/**
	 * No results have been attached to this node
	 *
	 * @param node node
	 */
	static public void setNoResult(@NonNull final TreeNode node)
	{
		// Log.d(TAG, "Deadend " + node);
		node.setDeadend(true);
		node.setEnabled(false);
	}

	public static void setTextNode(@NonNull final TreeNode node, final CharSequence text)
	{
		node.setValue(text);
	}

	public static void setTextNode(@NonNull final TreeNode node, @NonNull final CharSequence text, @DrawableRes final int icon)
	{
		node.setValue(new CompositeValue(text, icon, null, null));
	}
}

File: xNet/src/main/java/org/sqlunet/model/TreeFactory.kt
/*
 * Copyright (c) 2023. Bernard Bou
 */
package org.sqlunet.model

import android.os.Handler
import android.os.Looper
import androidx.annotation.DrawableRes
import org.sqlunet.treeview.control.ColdQueryTreeController
import org.sqlunet.treeview.control.HotQueryController
import org.sqlunet.treeview.control.HotQueryLinkController
import org.sqlunet.treeview.control.HotQueryTreeController
import org.sqlunet.treeview.control.IconTextController
import org.sqlunet.treeview.control.LeafController
import org.sqlunet.treeview.control.Link
import org.sqlunet.treeview.control.LinkController
import org.sqlunet.treeview.control.LinkHotQueryTreeController
import org.sqlunet.treeview.control.LinkLeafController
import org.sqlunet.treeview.control.LinkQueryTreeController
import org.sqlunet.treeview.control.LinkTreeController
import org.sqlunet.treeview.control.MoreController
import org.sqlunet.treeview.control.Query
import org.sqlunet.treeview.control.TextController
import org.sqlunet.treeview.control.TreeController
import org.sqlunet.treeview.model.TreeNode

/**
 * Tree factory
 *
 * @author [Bernard Bou](mailto:1313ou@gmail.com)
 */
object TreeFactory {
    // NON-TREE (without tree junction icon)

    /**
     * Make text node
     *
     * @param value       character sequence
     * @param breakExpand break expand flag
     * @return created node
     */
    fun makeTextNode(value: CharSequence, breakExpand: Boolean): TreeNode {
        return TreeNode(value, null, null, TextController(breakExpand), false)
    }

    /**
     * Make icon-text node
     *
     * @param text        text
     * @param icon        icon
     * @param breakExpand break expand flag
     * @return created node
     */
    fun makeIconTextNode(text: CharSequence, @DrawableRes icon: Int, breakExpand: Boolean): TreeNode {
        return TreeNode(text, icon, null, IconTextController(breakExpand), false)
    }

    /**
     * Make leaf node
     *
     * @param text        text
     * @param icon        icon (extra icon after tree icon)
     * @param breakExpand break expand flag
     * @return created node
     */
    fun makeLeafNode(text: CharSequence, @DrawableRes icon: Int, breakExpand: Boolean): TreeNode {
        return TreeNode(text, icon, null, LeafController(breakExpand), false)
    }

    /**
     * Make more (leaf) node
     *
     * @param text        text
     * @param icon        icon (extra icon after tree icon)
     * @param breakExpand break expand flag
     * @return created node
     */
    fun makeMoreNode(text: CharSequence, @DrawableRes icon: Int, breakExpand: Boolean): TreeNode {
        return TreeNode(text, icon, null, MoreController(breakExpand), false)
    }

    /**
     * Make icon-text-link node
     *
     * @param text        text
     * @param icon        icon
     * @param link        link
     * @param breakExpand break expand flag
     * @return created node
     */
    fun makeLinkNode(text: CharSequence, @DrawableRes icon: Int, breakExpand: Boolean, link: Link?): TreeNode {
        return TreeNode(text, icon, arrayOf(link), LinkController(breakExpand), false)
    }

    /**
     * Make link leaf node
     *
     * @param text        text
     * @param icon        icon
     * @param breakExpand break expand flag
     * @param link        link
     * @return created node
     */
    fun makeLinkLeafNode(text: CharSequence, @DrawableRes icon: Int, breakExpand: Boolean, link: Link?): TreeNode {
        return TreeNode(text, icon, arrayOf(link), LinkLeafController(breakExpand), false)
    }

    /**
     * Make hot (self-triggered) query node
     *
     * @param text        label text
     * @param icon        icon
     * @param breakExpand break expand flag
     * @param query       query
     * @return created node
     */
    fun makeHotQueryNode(text: CharSequence, @DrawableRes icon: Int, breakExpand: Boolean, query: Query?): TreeNode {
        val controller = HotQueryController(breakExpand)
        val result = TreeNode(text, icon, arrayOf(query, null), controller, false)
        val handler = Handler(Looper.getMainLooper())
        handler.post { controller.processQuery() }
        return result
    }

    /**
     * Make hot (self-triggered) query node
     *
     * @param text        label text
     * @param icon        icon
     * @param breakExpand break expand flag
     * @param query       query
     * @return created node
     */
    fun makeHotQueryLinkNode(text: CharSequence, @DrawableRes icon: Int, breakExpand: Boolean, query: Query?): TreeNode {
        val controller = HotQueryLinkController(breakExpand)
        val result = TreeNode(text, icon, arrayOf(query, null), controller, false)
        val handler = Handler(Looper.getMainLooper())
        handler.post { controller.processQuery() }
        return result
    }

    // TREE

    /**
     * Add tree node
     *
     * @param value       character sequence
     * @param icon        icon resource id
     * @param breakExpand break expand flag
     * @return created node
     */
    fun makeTreeNode(value: CharSequence, @DrawableRes icon: Int, breakExpand: Boolean): TreeNode {
        return TreeNode(value, icon, null, TreeController(breakExpand), true)
    }

    /**
     * Make link tree node
     *
     * @param text        text
     * @param icon        icon
     * @param breakExpand break expand flag
     * @param link        link
     * @return created node
     */
    fun makeLinkTreeNode(text: CharSequence, @DrawableRes icon: Int, breakExpand: Boolean, link: Link?): TreeNode {
        return TreeNode(text, icon, arrayOf(link), LinkTreeController(breakExpand), true)
    }

    /**
     * Make query tree node
     *
     * @param text        label text
     * @param icon        icon
     * @param breakExpand break expand flag
     * @param query       query
     * @return created node
     */
    fun makeQueryTreeNode(text: CharSequence, @DrawableRes icon: Int, breakExpand: Boolean, query: Query?): TreeNode {
        return TreeNode(text, icon, arrayOf(query), ColdQueryTreeController(breakExpand), true)
    }

    /**
     * Make hot (self-triggered) tree query node
     *
     * @param text        label text
     * @param icon        icon
     * @param breakExpand break expand flag
     * @param query       query
     * @return created node
     */
    fun makeHotQueryTreeNode(text: CharSequence, @DrawableRes icon: Int, breakExpand: Boolean, query: Query?): TreeNode {
        val controller = HotQueryTreeController(breakExpand)
        val result = TreeNode(text, icon, arrayOf(query), controller, true)
        val handler = Handler(Looper.getMainLooper())
        handler.post { controller.processQuery() }
        return result
    }

    /**
     * Make hot (self-triggered) tree query node
     *
     * @param text           label text
     * @param icon           icon
     * @param breakExpand    break expand flag
     * @param query          query
     * @param link           link
     * @param buttonImageRes image drawable id for button, 0 for default
     * @return created node
     */
    fun makeLinkHotQueryTreeNode(text: CharSequence, @DrawableRes icon: Int, breakExpand: Boolean, query: Query?, link: Link?, @DrawableRes buttonImageRes: Int): TreeNode {
        val controller: HotQueryTreeController = LinkHotQueryTreeController(breakExpand, buttonImageRes)
        val result = TreeNode(text, icon, arrayOf(query, link), controller, true)
        val handler = Handler(Looper.getMainLooper())
        handler.post { controller.processQuery() }
        return result
    }

    /**
     * Make link query tree node
     *
     * @param text           label text
     * @param icon           icon
     * @param breakExpand    break expand flag
     * @param query          query
     * @param link           link
     * @param buttonImageRes image drawable id for button, 0 for default
     * @return created node
     */
    fun makeLinkQueryTreeNode(text: CharSequence, @DrawableRes icon: Int, breakExpand: Boolean, query: Query?, link: Link?, @DrawableRes buttonImageRes: Int): TreeNode {
        return TreeNode(text, icon, arrayOf(query, link), LinkQueryTreeController(breakExpand, buttonImageRes), true)
    }

    // H E L P E R S

    /**
     * No results have been attached to this node
     *
     * @param node node
     */
    fun setNoResult(node: TreeNode) {
        // Log.d(TAG, "Deadend " + node)
        node.isDeadend = true
        node.isEnabled = false
    }

    fun setTextNode(node: TreeNode, text: CharSequence?) {
        node.text = text
    }

    fun setTextNode(node: TreeNode, text: CharSequence, @DrawableRes icon: Int) {
        node.text = text
        node.icon = icon
    }
}
